### DISCUSSION TITLE
Generate Code wo main code

### DISCUSSION THREAD

**Discussion by tomelgato** [davetcc/tcMenu#390] [Ideas]
Would be nice to be able to toggle between export of the main.cpp file or just a simple list with the callbacks. I dont like my main file to get messed up (my callbacks are in another file) and integrate the callbacks manually. Right now i need to delete the generated main file after every code generation....

**Comment by davetcc** [comment]
If you either type an @ in front of the function name or press edit next to the function edit box you can select declaration only. Either option does not create the callback instead just declaring the prototype in the project_menu.h file, then you must implement it somewhere yourself.

**Reply by tomelgato** [reply]
Where does TCMenu get the name from which it uses for he generated files? From the project folders name? (PIO environment). Would be nice to set custom names on project base, but i think we touched this subject already somewhere

The @ still doesnt solve the problem with creation of the main file, no? Like said it creates the main file which i need to delete every time (or i get a lot of compiler redeclaration errors). Not a biggie but a small inconvenience. Think i could put it also on ignore for the compiler, but didnt elaborate this further

**Reply by davetcc** [reply]
We'd need to look at options around that, the code that generates the main file is pretty isolated from everything else, so it could be done, really more a case of how to expose the functionality for this, I would imagine in the project properties that pop up when the root menu item is shown. 

The big thing around this is it would be an advanced option, but I even think the source generation folder is an advanced option as you can setup the project in a way that Arduino build can't handle. Maybe when this option is chosen a warning is needed, and during code generation a verification that there are no callback functions that require an implementation.

**Reply by davetcc** [reply]
So in terms of checking if there are any callback that require implementation, probably a stream operation here that throws if there are any matches, ie not starting with '@'.

https://github.com/davetcc/tcMenu/blob/d0ee186058485c6b544be98253b8bca82082233d/tcMenuGenerator/src/main/java/com/thecoderscorner/menu/editorui/generator/core/CoreCodeGenerator.java#L129 

and then here I guess we'd only generate/modify the main file if needed. 

https://github.com/davetcc/tcMenu/blob/d0ee186058485c6b544be98253b8bca82082233d/tcMenuGenerator/src/main/java/com/thecoderscorner/menu/editorui/generator/core/CoreCodeGenerator.java#L756C24-L756C24

**Reply by davetcc** [reply]
The only thing that would need to be sorted out is a little work in the list generation to ensure that every bit of code does not require a main implementation.

**Reply by davetcc** [reply]
> Where does TCMenu get the name from which it uses for he generated files?  

Project directory name for Arduino compatibility. The INO must match the project directory name.

**Reply by tomelgato** [reply]
True with INO but obsolete with PIO ;)

**Reply by tomelgato** [reply]
What about a checkbox which toggles between default INO environment (so like it is now) and custom environment.
With custom you are able to set the names for main and menu files and toggle if main file should be created at all

**Comment by davetcc** [comment]
This is possible using an `@` in front of the name, that effectively says, just export that function name, and assume that it will be implemented elsewhere. The web designer even has a function editor popup button next to the callback option.

For example

* `onCallback` - defined in main.
* `@onCallback` - exported only to be implemented elsewhere.
